quartic_validate/quartic_kl_plot.py

```diff
@@ -0,0 +1,256 @@
+
+            if (splitting not in destination):
+                destination[splitting] = (np.nan * np.zeros(len(timesteps)), np.nan * np.zeros(len(timesteps)))
+            destination[splitting][0][timestep_to_i[timestep]] = DF
+            destination[splitting][1][timestep_to_i[timestep]] = np.sqrt(dDF2)
+    return results_conf, results_full
+
+results_conf, results_full = get_results()
+
+# load the histograms and turn them into KL divergence arrays
+
+from pickle import dump
+with open("quartic_results/conf_results.pkl", "wb") as f:
+    dump(results_conf, f)
+
+with open("quartic_results/full_results.pkl", "wb") as f:
+    dump(results_full, f)
+
+# load numerical results
+conf_fname, full_fname = 'quartic_results/conf_results.pkl', 'quartic_results/full_results.pkl'
+with open(conf_fname, "rb") as f:
+    conf_results = load(f)
+with open(full_fname, "rb") as f:
+    full_results = load(f)
+
+# load noneq results
+fnames = glob('quartic_results/0*.npz')
+
+timesteps = []
+for fname in fnames:
+    file = np.load(fname)
+    dt, scheme = file['description']
+    timesteps.append(float(dt))
+timesteps = sorted(list(set(timesteps)))
+timestep_to_i = dict(zip(timesteps, range(len(timesteps))))
+
+colormaps = {
+    "RVOVR": "Greens",
+    "ORVRO": "Oranges",
+    "OVRVO": "Blues",
+    "VRORV": "Reds"
+}
+
+
+x = np.linspace(-10, 10, 10000)
+
+Z = np.trapz(np.exp(quartic.log_q(x)), x)
+
+def log_p(x):
+    return quartic.log_q(x) - np.log(Z)
+
+def kinetic_energy(v):
+    return 0.5 * quartic.mass * v**2
+
+mass = quartic.mass
+beta = quartic.beta
+velocity_scale = np.sqrt(1.0 / (beta * mass))
+sigma2 = quartic.velocity_scale**2
+potential = quartic.potential
+
+def log_v_density(v):
+    return -v ** 2 / (2 * sigma2) - np.log((np.sqrt(2 * np.pi * sigma2)))
+
+def p(x):
+    return np.exp(log_p(x))
+
+def v_density(v):
+    return np.exp(log_v_density(v))
+
+data_range = (-2.5, 2.5)
+n_bins = 50
+bin_edges = np.linspace(data_range[0], data_range[1], num=n_bins)
+
+one_d_hist_args = {"bins": n_bins * 2, "range": data_range, "density": True}
+one_d_bin_edges = np.linspace(data_range[0], data_range[1], num=n_bins*2 + 1)
+
+
+def normalize_histogram(hist, bin_edges):
+    x_range = bin_edges[-1] - bin_edges[0]
+    sum_y = np.sum(hist)
+    Z = (sum_y / x_range)
+    return hist / Z
+
+def compute_exact_histogram(density, bin_edges):
+    exact_hist = np.zeros(len(bin_edges) - 1)
+    for i in range(len(bin_edges) - 1):
+        left, right = bin_edges[i], bin_edges[i + 1]
+        x_ = np.linspace(left, right, 1000)
+        y_ = density(x_)
+        exact_hist[i] = np.trapz(y_, x_)
+    # let's double-check to make sure this histogram is normalized
+    return normalize_histogram(exact_hist, bin_edges)
+
+def compute_exact_joint_histogram(bin_edges):
+    x_eq_hist = compute_exact_histogram(p, bin_edges)
+    v_eq_hist = compute_exact_histogram(v_density, bin_edges)
+    joint_eq_hist = np.outer(v_eq_hist, x_eq_hist)
```

Comment:
Is this the right order for the outer product, so that `joint_eq_hist[i,j]` is `v_eq_hist[i] * x_eq_hist[j]`?